File: src/services/proactiveSuggestionService.ts
import { apiRequest } from './api';

// Tipos para sugerencias proactivas
export interface ProactiveSuggestion {
  id: number;
  user_id: number;
  suggestion_type: string;
  category?: string;
  priority: number;
  title: string;
  description?: string;
  reasoning?: string;
  expected_impact?: number;
  data_sources?: string[];
  confidence_score?: number;
  action_url?: string;
  action_type?: string;
  is_active: boolean;
  is_read: boolean;
  is_dismissed: boolean;
  is_applied: boolean;
  created_at: string;
  expires_at?: string;
  read_at?: string;
  dismissed_at?: string;
  applied_at?: string;
}

export interface SuggestionRule {
  id: number;
  rule_name: string;
  rule_type: string;
  conditions: any;
  suggestion_template: any;
  is_active: boolean;
  created_at: string;
}

export interface UserSuggestionPreference {
  id: number;
  user_id: number;
  notification_frequency: string;
  preferred_categories: string[];
  excluded_categories: string[];
  priority_threshold: number;
  max_suggestions_per_day: number;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

export interface SuggestionInsight {
  total_suggestions: number;
  read_suggestions: number;
  applied_suggestions: number;
  dismissed_suggestions: number;
  average_priority: number;
  most_effective_category: string;
  suggestion_effectiveness: number;
  generated_at: string;
}

// Servicio de sugerencias proactivas
export class ProactiveSuggestionService {
  private token: string;

  constructor(token: string) {
    this.token = token;
  }

  // Métodos estáticos para uso directo desde componentes
  private static getToken(): string {
    // Obtener token del localStorage o contexto de autenticación
    return localStorage.getItem('authToken') || '';
  }

  static async getDashboardSuggestions(): Promise<ProactiveSuggestion[]> {
    const token = ProactiveSuggestionService.getToken();
    if (!token) {
      throw new Error('No hay token de autenticación disponible');
    }
    
    const service = new ProactiveSuggestionService(token);
    return await service.getSuggestions(10);
  }

  static async markAsRead(suggestionId: number): Promise<{ success: boolean; message: string; data: { suggestion_id: number } }> {
    const token = ProactiveSuggestionService.getToken();
    if (!token) {
      throw new Error('No hay token de autenticación disponible');
    }
    
    const service = new ProactiveSuggestionService(token);
    return await service.markSuggestionRead(suggestionId);
  }

  static async dismissSuggestion(suggestionId: number): Promise<{ success: boolean; message: string; data: { suggestion_id: number } }> {
    const token = ProactiveSuggestionService.getToken();
    if (!token) {
      throw new Error('No hay token de autenticación disponible');
    }
    
    const service = new ProactiveSuggestionService(token);
    return await service.dismissSuggestion(suggestionId);
  }

  // Crear sugerencia
  async createSuggestion(suggestion: Omit<ProactiveSuggestion, 'id' | 'user_id' | 'created_at' | 'read_at' | 'dismissed_at' | 'applied_at'>): Promise<ProactiveSuggestion> {
    const response = await apiRequest<ProactiveSuggestion>(
      '/api/proactive-suggestions/',
      {
        method: 'POST',
        body: JSON.stringify(suggestion)
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al crear sugerencia');
    }

    return response.data!;
  }

  // Obtener sugerencias
  async getSuggestions(query: {
    suggestion_type?: string;
    category?: string;
    priority_min?: number;
    is_read?: boolean;
    is_dismissed?: boolean;
    limit?: number;
  } = {}): Promise<ProactiveSuggestion[]> {
    const params = new URLSearchParams();
    if (query.suggestion_type) params.append('suggestion_type', query.suggestion_type);
    if (query.category) params.append('category', query.category);
    if (query.priority_min) params.append('priority_min', query.priority_min.toString());
    if (query.is_read !== undefined) params.append('is_read', query.is_read.toString());
    if (query.is_dismissed !== undefined) params.append('is_dismissed', query.is_dismissed.toString());
    if (query.limit) params.append('limit', query.limit.toString());

    const response = await apiRequest<ProactiveSuggestion[]>(
      `/api/proactive-suggestions/?${params.toString()}`,
      {
        method: 'GET'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al obtener sugerencias');
    }

    return response.data!;
  }

  // Obtener sugerencias no leídas
  async getUnreadSuggestions(): Promise<ProactiveSuggestion[]> {
    const response = await apiRequest<ProactiveSuggestion[]>(
      '/api/proactive-suggestions/unread',
      {
        method: 'GET'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al obtener sugerencias no leídas');
    }

    return response.data!;
  }

  // Marcar sugerencia como leída
  async markAsRead(suggestionId: number): Promise<ProactiveSuggestion> {
    const response = await apiRequest<ProactiveSuggestion>(
      `/api/proactive-suggestions/${suggestionId}/read`,
      {
        method: 'PUT'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al marcar como leída');
    }

    return response.data!;
  }

  // Descartar sugerencia
  async dismissSuggestion(suggestionId: number): Promise<ProactiveSuggestion> {
    const response = await apiRequest<ProactiveSuggestion>(
      `/api/proactive-suggestions/${suggestionId}/dismiss`,
      {
        method: 'PUT'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al descartar sugerencia');
    }

    return response.data!;
  }

  // Obtener insights de sugerencias
  async getSuggestionInsights(): Promise<SuggestionInsight> {
    const response = await apiRequest<SuggestionInsight>(
      '/api/proactive-suggestions/insights',
      {
        method: 'GET'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al obtener insights');
    }

    return response.data!;
  }

  // Obtener preferencias del usuario
  async getUserPreferences(): Promise<UserSuggestionPreference> {
    const response = await apiRequest<UserSuggestionPreference>(
      '/api/proactive-suggestions/preferences',
      {
        method: 'GET'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al obtener preferencias');
    }

    return response.data!;
  }

  // Actualizar preferencias del usuario
  async updateUserPreferences(preferences: Partial<UserSuggestionPreference>): Promise<UserSuggestionPreference> {
    const response = await apiRequest<UserSuggestionPreference>(
      '/api/proactive-suggestions/preferences',
      {
        method: 'PUT',
        body: JSON.stringify(preferences)
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al actualizar preferencias');
    }

    return response.data!;
  }

  // Generar sugerencias de proyecto
  async generateProjectSuggestions(): Promise<ProactiveSuggestion[]> {
    const response = await apiRequest<ProactiveSuggestion[]>(
      '/api/proactive-suggestions/generate/project',
      {
        method: 'POST'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al generar sugerencias de proyecto');
    }

    return response.data!;
  }

  // Generar sugerencias de aprendizaje
  async generateLearningSuggestions(): Promise<ProactiveSuggestion[]> {
    const response = await apiRequest<ProactiveSuggestion[]>(
      '/api/proactive-suggestions/generate/learning',
      {
        method: 'POST'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al generar sugerencias de aprendizaje');
    }

    return response.data!;
  }

  // Generar sugerencias basadas en tiempo
  async generateTimeBasedSuggestions(): Promise<ProactiveSuggestion[]> {
    const response = await apiRequest<ProactiveSuggestion[]>(
      '/api/proactive-suggestions/generate/time-based',
      {
        method: 'POST'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al generar sugerencias basadas en tiempo');
    }

    return response.data!;
  }

  // Generar todas las sugerencias
  async generateAllSuggestions(): Promise<ProactiveSuggestion[]> {
    const response = await apiRequest<ProactiveSuggestion[]>(
      '/api/proactive-suggestions/generate/all',
      {
        method: 'POST'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al generar todas las sugerencias');
    }

    return response.data!;
  }

  // Limpiar sugerencias expiradas
  async cleanupExpiredSuggestions(): Promise<{ message: string }> {
    const response = await apiRequest<{ message: string }>(
      '/api/proactive-suggestions/cleanup',
      {
        method: 'POST'
      },
      this.token
    );

    if (response.status === 'error') {
      throw new Error(response.error || 'Error al limpiar sugerencias');
    }

    return response.data!;
  }
}

File: backend/app/services/proactive_suggestion_service.py
import json
from datetime import datetime, timedelta
from typing import Dict, Any, List, Optional
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, desc, func
from ..models.proactive_suggestions import (
    ProactiveSuggestion, 
    SuggestionRule, 
    UserSuggestionPreference
)
from ..models.project import Project
# from ..models.project_activity import ProjectActivity
from ..models.business_model_canvas import BusinessModelCanvas
from ..models.agent_memory import UserContext
from ..schemas.proactive_suggestions import (
    ProactiveSuggestionCreate,
    ProactiveSuggestionUpdate,
    SuggestionRuleCreate,
    SuggestionRuleUpdate,
    UserSuggestionPreferenceCreate,
    UserSuggestionPreferenceUpdate,
    SuggestionQuery,
    SuggestionInsight
)


class ProactiveSuggestionService:
    def __init__(self, db: Session):
        self.db = db
    
    def create_suggestion(self, user_id: int, suggestion_data: ProactiveSuggestionCreate) -> ProactiveSuggestion:
        """Crea una nueva sugerencia proactiva"""
        suggestion = ProactiveSuggestion(
            user_id=user_id,
            **suggestion_data.dict()
        )
        self.db.add(suggestion)
        self.db.commit()
        self.db.refresh(suggestion)
        return suggestion
    
    def get_suggestions(self, user_id: int, query: SuggestionQuery) -> List[ProactiveSuggestion]:
        """Obtiene sugerencias según criterios de búsqueda"""
        db_query = self.db.query(ProactiveSuggestion).filter(ProactiveSuggestion.user_id == user_id)
        
        if query.suggestion_type:
            db_query = db_query.filter(ProactiveSuggestion.suggestion_type == query.suggestion_type)
        
        if query.category:
            db_query = db_query.filter(ProactiveSuggestion.category == query.category)
        
        if query.priority_min:
            db_query = db_query.filter(ProactiveSuggestion.priority >= query.priority_min)
        
        if query.is_read is not None:
            db_query = db_query.filter(ProactiveSuggestion.is_read == query.is_read)
        
        if query.is_dismissed is not None:
            db_query = db_query.filter(ProactiveSuggestion.is_dismissed == query.is_dismissed)
        
        return db_query.order_by(desc(ProactiveSuggestion.priority), desc(ProactiveSuggestion.created_at)).limit(query.limit).all()
    
    def mark_as_read(self, suggestion_id: int, user_id: int) -> Optional[ProactiveSuggestion]:
        """Marca una sugerencia como leída"""
        suggestion = self.db.query(ProactiveSuggestion).filter(
            and_(
                ProactiveSuggestion.id == suggestion_id,
                ProactiveSuggestion.user_id == user_id
            )
        ).first()
        
        if not suggestion:
            return None
        
        suggestion.is_read = True
        suggestion.read_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(suggestion)
        return suggestion
    
    def dismiss_suggestion(self, suggestion_id: int, user_id: int) -> Optional[ProactiveSuggestion]:
        """Descarta una sugerencia"""
        suggestion = self.db.query(ProactiveSuggestion).filter(
            and_(
                ProactiveSuggestion.id == suggestion_id,
                ProactiveSuggestion.user_id == user_id
            )
        ).first()
        
        if not suggestion:
            return None
        
        suggestion.is_dismissed = True
        suggestion.dismissed_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(suggestion)
        return suggestion
    
    def get_or_create_user_preferences(self, user_id: int) -> UserSuggestionPreference:
        """Obtiene o crea las preferencias del usuario"""
        preferences = self.db.query(UserSuggestionPreference).filter(
            UserSuggestionPreference.user_id == user_id
        ).first()
        
        if not preferences:
            preferences = UserSuggestionPreference(user_id=user_id)
            self.db.add(preferences)
            self.db.commit()
            self.db.refresh(preferences)
        
        return preferences
    
    def update_user_preferences(self, user_id: int, update_data: UserSuggestionPreferenceUpdate) -> UserSuggestionPreference:
        """Actualiza las preferencias del usuario"""
        preferences = self.get_or_create_user_preferences(user_id)
        
        for field, value in update_data.dict(exclude_unset=True).items():
            setattr(preferences, field, value)
        
        preferences.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(preferences)
        return preferences
    
    def get_suggestion_insights(self, user_id: int) -> SuggestionInsight:
        """Obtiene insights sobre las sugerencias del usuario"""
        suggestions = self.db.query(ProactiveSuggestion).filter(
            ProactiveSuggestion.user_id == user_id
        ).all()
        
        total_suggestions = len(suggestions)
        unread_suggestions = len([s for s in suggestions if not s.is_read])
        high_priority_suggestions = len([s for s in suggestions if s.priority >= 7])
        
        suggestions_by_category = {}
        suggestions_by_type = {}
        total_priority = 0
        
        for suggestion in suggestions:
            # Por categoría
            category = suggestion.category or "general"
            suggestions_by_category[category] = suggestions_by_category.get(category, 0) + 1
            
            # Por tipo
            suggestions_by_type[suggestion.suggestion_type] = suggestions_by_type.get(suggestion.suggestion_type, 0) + 1
            
            total_priority += suggestion.priority
        
        average_priority = total_priority / total_suggestions if total_suggestions > 0 else 0
        
        return SuggestionInsight(
            total_suggestions=total_suggestions,
            unread_suggestions=unread_suggestions,
            high_priority_suggestions=high_priority_suggestions,
            suggestions_by_category=suggestions_by_category,
            suggestions_by_type=suggestions_by_type,
            average_priority=average_priority
        )
    
    def generate_project_suggestions(self, user_id: int) -> List[ProactiveSuggestion]:
        """Genera sugerencias basadas en el estado de los proyectos"""
        suggestions = []
        
        # Obtener proyectos del usuario
        projects = self.db.query(Project).filter(Project.owner_id == user_id).all()
        
        for project in projects:
            # Sugerencia para proyectos sin BMC
            bmc = self.db.query(BusinessModelCanvas).filter(
                BusinessModelCanvas.project_id == project.id
            ).first()
            
            if not bmc:
                suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                    suggestion_type="recommendation",
                    title="Generar Business Model Canvas",
                    description=f"Tu proyecto '{project.name}' aún no tiene un Business Model Canvas. Te sugiero generarlo para tener una visión clara de tu modelo de negocio.",
                    priority=8,
                    category="bmc",
                    action_url=f"/projects/{project.id}",
                    action_text="Generar BMC",
                    context_data={"project_id": project.id, "project_name": project.name}
                )))
            
            # Sugerencia para proyectos sin actividades
            activities = self.db.query(ProjectActivity).filter(
                ProjectActivity.project_id == project.id
            ).all()
            
            if not activities:
                suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                    suggestion_type="recommendation",
                    title="Crear actividades para el proyecto",
                    description=f"Tu proyecto '{project.name}' no tiene actividades definidas. Te sugiero crear un plan de acción con tareas específicas.",
                    priority=7,
                    category="activity",
                    action_url=f"/projects/{project.id}/activities",
                    action_text="Crear Actividades",
                    context_data={"project_id": project.id, "project_name": project.name}
                )))
            
            # Sugerencia para proyectos inactivos
            if project.status == "inactive" and (datetime.utcnow() - project.updated_at).days > 7:
                suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                    suggestion_type="reminder",
                    title="Proyecto inactivo",
                    description=f"Tu proyecto '{project.name}' lleva más de una semana sin actividad. ¿Te gustaría revisarlo o actualizar su estado?",
                    priority=6,
                    category="project",
                    action_url=f"/projects/{project.id}",
                    action_text="Revisar Proyecto",
                    context_data={"project_id": project.id, "project_name": project.name}
                )))
        
        return suggestions
    
    def generate_learning_suggestions(self, user_id: int) -> List[ProactiveSuggestion]:
        """Genera sugerencias de aprendizaje basadas en el contexto del usuario"""
        suggestions = []
        
        # Obtener contexto del usuario
        user_context = self.db.query(UserContext).filter(UserContext.user_id == user_id).first()
        
        if user_context:
            expertise_level = user_context.expertise_level
            business_sector = user_context.business_sector
            
            # Sugerencias basadas en nivel de experiencia
            if expertise_level == "beginner":
                suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                    suggestion_type="recommendation",
                    title="Aprende sobre Business Model Canvas",
                    description="Como emprendedor principiante, te recomiendo aprender sobre el Business Model Canvas para estructurar mejor tu idea de negocio.",
                    priority=6,
                    category="learning",
                    action_url="/learn",
                    action_text="Ir a Aprende",
                    context_data={"topic": "bmc", "level": "beginner"}
                )))
            
            # Sugerencias basadas en sector de negocio
            if business_sector:
                suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                    suggestion_type="opportunity",
                    title=f"Tendencias en {business_sector}",
                    description=f"Descubre las últimas tendencias y oportunidades en el sector {business_sector} para mantener tu proyecto actualizado.",
                    priority=5,
                    category="learning",
                    action_url="/learn",
                    action_text="Ver Tendencias",
                    context_data={"sector": business_sector}
                )))
        
        return suggestions
    
    def generate_time_based_suggestions(self, user_id: int) -> List[ProactiveSuggestion]:
        """Genera sugerencias basadas en el tiempo y patrones de uso"""
        suggestions = []
        
        # Obtener preferencias del usuario
        preferences = self.get_or_create_user_preferences(user_id)
        
        # Sugerencia de revisión semanal
        last_suggestion = self.db.query(ProactiveSuggestion).filter(
            and_(
                ProactiveSuggestion.user_id == user_id,
                ProactiveSuggestion.suggestion_type == "reminder",
                ProactiveSuggestion.category == "weekly_review"
            )
        ).order_by(desc(ProactiveSuggestion.created_at)).first()
        
        if not last_suggestion or (datetime.utcnow() - last_suggestion.created_at).days >= 7:
            suggestions.append(self.create_suggestion(user_id, ProactiveSuggestionCreate(
                suggestion_type="reminder",
                title="Revisión Semanal de Proyectos",
                description="Es momento de revisar el progreso de tus proyectos. Evalúa lo que has logrado y planifica las próximas acciones.",
                priority=5,
                category="weekly_review",
                action_url="/dashboard",
                action_text="Revisar Proyectos",
                context_data={"type": "weekly_review"}
            )))
        
        return suggestions
    
    def cleanup_expired_suggestions(self) -> int:
        """Limpia sugerencias expiradas"""
        expired_suggestions = self.db.query(ProactiveSuggestion).filter(
            and_(
                ProactiveSuggestion.expires_at.isnot(None),
                ProactiveSuggestion.expires_at <= datetime.utcnow()
            )
        ).all()
        
        count = len(expired_suggestions)
        for suggestion in expired_suggestions:
            self.db.delete(suggestion)
        
        self.db.commit()
        return count
